### PR TITLE
[sbt 1.0] Reformat using Scalafmt

### LINKED ISSUE #2944: clean up source code
So, at least for me the state of the source code is a big barrier to contributing. It's poorly formatted, sparsely commented, and in places it's organized strangely. So I suggest a cleanup task to be done at some point post-1.0 when we're free to make "big" changes. In this case the changes are trivial but global and will result in a huge PR. Anyway:

- Run everything through scalafmt to make formatting consistent and introduce some whitespace between definitions.
- Add Scaladoc comments explaining at least what every class does. Ideally every public member would be documented but that's too much to ask initially.
- In some places there are a bunch of definitions in the same source file that might work better split up into multiple source files.

After this is done I think it will be easier for me (and hopefully others) to get some momentum in trying to understand how the hell sbt actually works so we can start making meaningful improvements.

### PR DESCRIPTION
**Pull request by eed3si9n**
Fixes #2944

Taking recommendation by @olafurpg in #3003:

> - maxColumn 100 is totally fine.
> - don't use vertical alignment, it's distracting.
> - use default or I'm happy to discuss adding style=sbt, sbt is a large enough project to merit its own style.

The 4 tweakings I took from original @dwijnand PRs with my added comments:

```yaml

# http://docs.scala-lang.org/style/scaladoc.html recommends the JavaDoc style.
# scala/scala is written that way too https://github.com/scala/scala/blob/v2.12.2/src/library/scala/Predef.scala
docstrings = JavaDoc

# This also seems more idiomatic to include whitespace in import x.{ yyy }
spaces.inImportCurlyBraces = true

# This works around sequence wildcard (`_*`) turning into `_ *`
spaces.beforeSeqWildcard = true

# Vertical alignment only => for pattern matching
align.tokens.add = [
  { code = "=>",  owner = "Case"  }
]
```

Note this adds automatic formatting on `compile` task using `AutomateScalafmtPlugin` implemented by @hseeberger, and CI checking.

### CODE REVIEW

File: main-actions/src/main/scala/sbt/ForkTests.scala
Comment (smarter) [inline main-actions/src/main/scala/sbt/ForkTests.scala:127]: This is a pretty weird formatting.

**Comment by jvican** [comment]
Can we please merge this after my PR is merged? Rebasing it on top of this will be very difficult and I prefer to spend that time on other things.

File: main-actions/src/main/scala/sbt/Sync.scala
Comment (smarter) [inline main-actions/src/main/scala/sbt/Sync.scala:63]: This is also weird, you want `else if (...) {` on a single line, not indenting the block as if it was a normal expression.

File: main-actions/src/test/scala/sbt/compiler/EvalTest.scala
Comment (smarter) [inline main-actions/src/test/scala/sbt/compiler/EvalTest.scala:81]: Also unlike any kind of style a human would use.

**Comment by smarter** [comment]
@olafurpg Does scalafmt take operator precedence into account when deciding where to put newlines?

File: main-command/src/main/scala/sbt/State.scala
Comment (dwijnand) [inline main-command/src/main/scala/sbt/State.scala:136]: holy shit when did scalafmt start doing this newline inserting? <3 @olafurpg

File: main-settings/src/main/scala/sbt/Structure.scala
Comment (dwijnand) [inline main-settings/src/main/scala/sbt/Structure.scala]: jesus christ. We should exclude this section and the same one above. They're exceptions.

File: main-actions/src/main/scala/sbt/Package.scala
Comment (eed3si9n) [inline main-actions/src/main/scala/sbt/Package.scala]: Changing `_*` to `_ *` looks odd.  

SLS says "The last pattern in a pattern sequence may be a sequence wildcard _*." - https://www.scala-lang.org/files/archive/spec/2.12/08-pattern-matching.html#pattern-sequences

**Comment by tpolecat** [comment]
The lack of blank lines between definitions is among my top complaints and this doesn't seem to affect it. See [Eval.scala](https://github.com/eed3si9n/sbt/blob/499bba8142a988ffec713bd9db0db1c117c97dd7/main-actions/src/main/scala/sbt/compiler/Eval.scala) for example. Is there an easy way to improve that?

Otherwise the formatting isn't really my taste but it's consistent and it's an improvement so 👍 from me. I'm kind of ambivalent about enforcing it as part of the build vs. just doing it once and calling it good. No strong opinion.

File: main/src/main/scala/sbt/Keys.scala
Comment (dwijnand) [inline main/src/main/scala/sbt/Keys.scala]: uh... do we want to change this like this? I know the former is longer than 100 columns, but I thinks it's about 1000x more readable. We should exclude these blocks, maybe the whole file.

File: main/src/main/scala/sbt/Main.scala
Comment (dwijnand) [inline main/src/main/scala/sbt/Main.scala:219]: these binpacked would be much nicer..

File: .scalafmt.conf
Comment (dwijnand) [inline .scalafmt.conf]: we should exclude `/contraband-scala/` too otherwise scalafmt and contraband are going to keep fighting each other

File: main-actions/src/main/scala/sbt/Package.scala
Comment (dwijnand) [reply main-actions/src/main/scala/sbt/Package.scala]: It's why I'd set `spaces.beforeSeqWildcard = true`. I've no idea why the default is false. scalameta/scalafmt#772

File: main-actions/src/main/scala/sbt/Package.scala
Comment (dwijnand) [inline main-actions/src/main/scala/sbt/Package.scala]: Can we at least agree that `=>`s should be aligned? I find the former state to be so much more legible..

File: project/Formatting.scala
Comment (dwijnand) [inline project/Formatting.scala:37]: We should continue to enforce the format with CI.

**Review by dwijnand — CHANGES_REQUESTED** [review]
Thank you for the PR, @eed3si9n. Looks generally good, I've left a few comments to discuss.

File: main/src/main/scala/sbt/Keys.scala
Comment (eed3si9n) [reply main/src/main/scala/sbt/Keys.scala]: Maybe we need a Perl script to lift the key description such as `"The amount of logging sent to a file for persistence."` into a variable name right above the `val`.

File: main-actions/src/main/scala/sbt/Package.scala
Comment (eed3si9n) [reply main-actions/src/main/scala/sbt/Package.scala]: Sure why not.

File: main-settings/src/main/scala/sbt/Structure.scala
Comment (dwijnand) [inline main-settings/src/main/scala/sbt/Structure.scala]: The way to exclude this block is by surrounding it with

```
// format: off
```

and then

```
// format: on
```

same for the t2ToApp2, etc block below.

I can do it in a follow up PR if you'd prefer.

**Review by dwijnand — APPROVED** [review]

File: main-actions/src/main/scala/sbt/ForkTests.scala
Comment (olafurpg) [reply main-actions/src/main/scala/sbt/ForkTests.scala:127]: Yes, I agree. Scalafmt leaves respects line breaks around infix operators. If you put a newline between `+` and `options` the formatting would look nicer.

File: main-actions/src/test/scala/sbt/compiler/EvalTest.scala
Comment (olafurpg) [reply main-actions/src/test/scala/sbt/compiler/EvalTest.scala:81]: This is a known issue https://github.com/scalameta/scalafmt/issues/667 I suspect it would look nicer wrapped in curly braces.

File: main-command/src/main/scala/sbt/State.scala
Comment (olafurpg) [reply main-command/src/main/scala/sbt/State.scala:136]: Scalafmt inserts a blank line if there's a docstring :))

File: main-actions/src/main/scala/sbt/Package.scala
Comment (olafurpg) [reply main-actions/src/main/scala/sbt/Package.scala]: I've opened https://github.com/scalameta/scalafmt/issues/890

**Comment by olafurpg** [comment]
@tpolecat There is a `newlines.alwaysBeforeTopLevelStatements` setting http://scalameta.org/scalafmt/#newlines.alwaysBeforeTopLevelStatements which was contributed by @Daxten to enforce blank lines between definitions. Currently, scalafmt only enforces a blank before docstrings.

File: main-actions/src/main/scala/sbt/Package.scala
Comment (olafurpg) [reply main-actions/src/main/scala/sbt/Package.scala]: FWIW, the Scala.js coding style permits vertical alignment in pattern matching only.

You can limit vertical alignment to `=>` in pattern matching only with 

```
align.tokens = [{owner="Case", code="=>"}]
```

That will avoid unwanted alignment for lambda arrows, import rename arrows etc.

File: main-actions/src/test/scala/sbt/compiler/EvalTest.scala
Comment (olafurpg) [reply main-actions/src/test/scala/sbt/compiler/EvalTest.scala:81]: Wrapping it in curly braces and inserting a newline after `==` yields this output

```scala
  private[this] def testImport(imports: Seq[String]): Int => Prop = { i =>
    value(eval.eval("abs(" + i + ")", new EvalImports(imports.zipWithIndex, "imp"))) ==
      math.abs(i)
  }
```

I tried really hard to come up with a better solution for infix operators but it always came out badly in some weird DSLs in the wild. Until scalafmt is able to handle infix operators well in all cases I thought it was best to leave those line break choices to the user.

**Comment by olafurpg** [comment]
First of all, I'm really happy to see this PR 😄 A few notes

- scalafmt does not insert or remove line breaks around infix operators, so weird formatting like below can usually be fixed with a bit of manual help

```scala
foo(aaaaaa(bbbbbbbbbbbbbbbb), 333333333333) == a(
  2)
// manually insert newline after `==`
foo(aaaaaa(bbbbbbbbbbbbbbbb), 333333333333) ==
  a(2)
```
I really wanted scalafmt to be able handle infix operators well (at least common ones like == && ||) but I always hit on places where I didn't like the output. Instead of forcing bad formatted output without an escape hatch, scalafmt respects the line break choices in the original source for infix operators.

- if you enforce scalafmt in CI with `scalafmt --test`, you can customize the failure message with to help contributors figure out how to fix the problem.

```scala
onTestFailure = "To fix this error, run XXX".
```

- if you really dislike the output, please report an issue with the original code, formatted code and your expected output. Some things are super easy to fix while others are a lot harder. 

- the `docstring`/`spaces.inImportCurlyBraces`/`spaces.beforeSeqWildcard` options are totally reasonable. I've opened https://github.com/scalameta/scalafmt/issues/890 to discuss enabling those in the default style. If that meets big resistance, I'm open to add `style=sbt`.

Finally, my experience is that the formatted output tends to look a bit nicer when new code is written with scalafmt in mind (esp. max column) compared to when an existing codebase is formatted with scalafmt.

**Comment by dwijnand** [comment]
@tpolecat @olafurpg I too find spaces between multi-line definitions to make code much better. The reason I didn't include `newlines.alwaysBeforeTopLevelStatements=true` was because of issues in how that works. I've taken the time to file those issues now (https://github.com/scalameta/scalafmt/issues/892, https://github.com/scalameta/scalafmt/issues/893), then we can do that formatting.

**Comment by Daxten** [comment]
Will take a look this weekend!

**Comment by SethTisue** [comment]
scalafmt all the things! 🎉

File: project/AutomateScalafmtPlugin.scala
Comment (eed3si9n) [inline project/AutomateScalafmtPlugin.scala:61]: @dwijnand See here. It requires ScalafmtPlugin.

**Pull request merged**
